containers/pre_built_workers/csharp/Dockerfile

```diff
@@ -14,26 +14,36 @@
 
 FROM mcr.microsoft.com/dotnet/sdk:2.1
 
+# TODO: when running on kokoro, this build step will not be cached
+# since we'll always be on a fresh VM. Re-running this command each
+# time leads to increased latency and flakiness.
+RUN apt-get update && apt-get install -y \
+  build-essential \
+  git \
+  cmake && \
+  apt-get clean
+
 RUN mkdir -p /pre
 WORKDIR /pre
 
 ARG REPOSITORY=grpc/grpc
 ARG GITREF=master
+# when BREAK_CACHE arg is set to a random value (e.g. by "--build-arg BREAK_CACHE=$(uuidgen)"),
+# it makes sure the docker cache breaks at this command, and all the following
+# commands in this Dockerfile will be forced to re-run on each build.
+# This is important to ensure we always clone the repository even if "GITREF" stays unchanged
+# (important e.g. when GITREF=master, when the clone command could get cached and
+# we'd end up with a stale repository).
+ARG BREAK_CACHE
 
 RUN git clone https://github.com/$REPOSITORY.git .
-RUN git submodule update --init
 RUN git checkout $GITREF
+RUN git submodule update --init
```

Comment:
I can fix that after we merge this PR if the fetch method makes more scene.

Reply:
So the original problem was different:
`git submodule update --init` needs to come AFTER the checkout, otherwise you're updating the submodules to a different version (if HEAD has different version of submodules than you're desired GITREF, you'll end up with submodules that correspond to the head, not to your gitref - and that will lead to very odd build errors that can be hard to debug).

Besides that:
- the approach you mentioned (start with `git init`) is actually identical in terms of how much data needs to be fetched from github (so same latency and same chance of adding flakiness). The advantage is that it will allow you to clone even some very old versions of the repository where the layout of submodule is very different (sometimes what used to be a submodule is now a directory in the main repo or vice versa) and that can cause trouble when you first checkout master branch and only after that you try to do git checkout $GITREF. @codeblooded has ran into this (e.g. see here: https://github.com/grpc/test-infra/pull/86).
- `git submodule update --init` is sufficient for the grpc/grpc repository and `git submodule update --init --recursive` downloads submodule for our submodules (which AFAIK is unnecessary and wastes bandwidth and creates opportunity for download flakes) - so just using `git submodule update --init` is better from that perspective.

Ad "cloning the whole repository multiple times":
What I had in mind is that under real usage, you'd have a clone of grpc/grpc of the right version already cloned locally. But then you basically fetch the grpc repo with all submodules multiple times (once for C++, once for csharp, once for ruby, once for php etc.). In the ideal world, this is something we'd avoid, but since doing that as part of docker build is tricky, I'd leave this as is for now. I just wanted to point out the limitation of the current approach.